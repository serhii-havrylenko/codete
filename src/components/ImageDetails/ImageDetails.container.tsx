import { inject, observer } from 'mobx-react';
import * as React from 'react';

import { Stores } from '../../stores';
import Images, { ImageDetailsProps } from './ImageDetails';

const ImagesDetailsContainer = inject<
  Stores,
  {},
  Pick<ImageDetailsProps, 'image' | 'onDeleteClick' | 'onEditClick'>,
  {}
>(({ imageDetails: imageDetailsStore }: Stores) => {
  return {
    image: imageDetailsStore.image,
    onDeleteClick: console.log,
    onEditClick: console.log,
  };
})(observer(Images as React.FunctionComponent<{}>));

export default ImagesDetailsContainer;
