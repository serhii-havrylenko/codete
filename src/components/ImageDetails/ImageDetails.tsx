import Avatar from '@material-ui/core/Avatar';
import Button from '@material-ui/core/Button';
import Card from '@material-ui/core/Card';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import CardHeader from '@material-ui/core/CardHeader';
import CardMedia from '@material-ui/core/CardMedia';
import red from '@material-ui/core/colors/red';
import IconButton from '@material-ui/core/IconButton';
import { StyleRulesCallback, withStyles } from '@material-ui/core/styles';
import { ClassNameMap } from '@material-ui/core/styles/withStyles';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';
import CancelIcon from '@material-ui/icons/Cancel';
import DeleteIcon from '@material-ui/icons/Delete';
import EditIcon from '@material-ui/icons/Edit';
import SaveIcon from '@material-ui/icons/Save';
import * as React from 'react';

import { Image } from '../../types';

export interface ImageDetailsProps {
  image: Image | null;
  classes: ClassNameMap;
  onDeleteClick: () => void;
  onEditClick: (title: Image['title']) => void;
}

interface ImageDetailsState {
  edit: boolean;
  titleHasError: boolean;
}

const styles: StyleRulesCallback = (theme) => ({
  card: {
    maxWidth: '700px',
  },
  media: {
    objectFit: 'cover',
  },
  actions: {
    display: 'flex',
    justifyContent: 'flex-end',
  },
  avatar: {
    backgroundColor: red[500],
  },
});

const CardHeaderStyled = withStyles({
  action: {
    alignSelf: 'center',
  },
})(CardHeader);

class ImageDetails extends React.Component<
  ImageDetailsProps,
  ImageDetailsState
> {
  public state = {
    edit: false,
    titleHasError: false,
  };

  private titleRef = React.createRef<HTMLInputElement>();

  public render() {
    const { classes, image } = this.props;
    const { edit, titleHasError } = this.state;

    if (!image) {
      return <div>Nothing to show. Image is not found</div>;
    }

    return (
      <Card className={classes.card}>
        <CardHeaderStyled
          avatar={
            <Avatar className={classes.avatar}>
              {image.author.substring(0, 4)}
            </Avatar>
          }
          action={
            edit ? (
              <div>
                <IconButton onClick={this.handleOnSaveEditClick}>
                  <SaveIcon />
                </IconButton>
                <IconButton onClick={this.handleOnCancelEditClick}>
                  <CancelIcon />
                </IconButton>
              </div>
            ) : (
              <IconButton onClick={this.handleOnEditClick}>
                <EditIcon />
              </IconButton>
            )
          }
          title={
            edit ? (
              <div>
                <TextField
                  margin="none"
                  inputRef={this.titleRef}
                  label="Title"
                  required={true}
                  type="text"
                  error={titleHasError}
                  helperText={titleHasError && 'This field is required'}
                />
              </div>
            ) : (
              image.title
            )
          }
          subheader={new Date(image.created).toLocaleString()}
        />
        <CardMedia
          className={classes.media}
          src={image.src}
          component="img"
          title={image.title}
        />
        <CardContent>
          <Typography component="pre">{image.details}</Typography>
        </CardContent>
        <CardActions className={classes.actions} disableActionSpacing={true}>
          <Button onClick={this.handleOnDeleteClick}>
            <DeleteIcon />
            Delete
          </Button>
        </CardActions>
      </Card>
    );
  }

  private handleOnDeleteClick = () => {
    this.props.onDeleteClick();
  };

  private handleOnEditClick = () => {
    this.setState({ edit: true }, () => {
      if (this.titleRef.current) {
        this.titleRef.current.focus();
        this.titleRef.current.value = this.props.image!.title;
      }
    });
  };

  private handleOnSaveEditClick = () => {
    if (!this.titleRef.current) {
      return;
    }

    const value = this.titleRef.current.value;

    if (!value) {
      this.setState({ titleHasError: true });
      return;
    }

    this.setState({ edit: false, titleHasError: false });
    this.props.onEditClick(value);
  };

  private handleOnCancelEditClick = () => {
    this.setState({ edit: false, titleHasError: false });
  };
}

export default withStyles(styles)(ImageDetails);
