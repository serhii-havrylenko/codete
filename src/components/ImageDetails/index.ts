export { default } from './ImageDetails';

export { default as ImageDetailsContainer } from './ImageDetails.container';
