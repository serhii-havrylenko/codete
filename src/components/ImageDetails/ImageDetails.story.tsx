import { action } from '@storybook/addon-actions';
import { text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';

import ImageDetails from './ImageDetails';

storiesOf('ImageDetails', module)
  .add('default', () => {
    const props = {
      image: {
        id: '1',
        title: text('Image 1 text', 'Image 1'),
        created: 1555853943382,
        details: text('Image 1 details', 'details\nnext line'),
        author: 'Rob',
        src: text('Image 1 src', 'https://via.placeholder.com/150'),
      },
      onDeleteClick: action('Delete clicked'),
      onEditClick: action('Edit clicked'),
    };

    return <ImageDetails {...props} />;
  })
  .add('without image', () => {
    const props = {
      onDeleteClick: action('Delete clicked'),
      onEditClick: action('Edit clicked'),
      image: null,
    };
    return <ImageDetails {...props} />;
  });
