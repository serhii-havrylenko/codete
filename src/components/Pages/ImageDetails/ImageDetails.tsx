import { inject } from 'mobx-react';
import * as React from 'react';
import { match } from 'react-router';

import { Stores } from '../../../stores';
import { ImageDetailsContainer } from '../../ImageDetails';

interface ImageDetailsProps {
  match: match<{ id: string }>;
  imageDetails: Stores['imageDetails'];
}

@inject('imageDetails')
class ImageDetails extends React.Component<ImageDetailsProps> {
  constructor(props: ImageDetailsProps) {
    super(props);

    props.imageDetails.setActiveImage(props.match.params.id);
  }

  public render() {
    return <ImageDetailsContainer />;
  }
}

export default ImageDetails;
