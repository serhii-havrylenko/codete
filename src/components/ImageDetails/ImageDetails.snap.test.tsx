import { mount } from 'enzyme';
import * as React from 'react';

import { Theme } from '@material-ui/core';
import ImageDetails, { styles } from './ImageDetails';

jest.mock('@material-ui/core/Card', () => ({ children, ...rest }: any) => (
  <div {...rest}>{children}</div>
));
jest.mock('@material-ui/core/CardActions', () => ({ children }: any) => (
  <div>{children}</div>
));
jest.mock(
  '@material-ui/core/CardContent',
  () => ({ children, ...rest }: any) => <div {...rest}>{children}</div>,
);
jest.mock(
  '@material-ui/core/CardHeader',
  () => ({ children, ...rest }: any) => <div {...rest}>{children}</div>,
);
jest.mock('@material-ui/core/CardMedia', () => ({ children }: any) => (
  <div>{children}</div>
));
jest.mock('@material-ui/core/Avatar', () => ({ children }: any) => (
  <div>{children}</div>
));
jest.mock('@material-ui/core/Button', () => ({ children, ...rest }: any) => (
  <button {...rest}>{children}</button>
));
jest.mock(
  '@material-ui/core/IconButton',
  () => ({ children, ...rest }: any) => <button {...rest}>{children}</button>,
);
jest.mock('@material-ui/core/TextField', () => ({ children, ...rest }: any) => (
  <input {...rest}>{children}</input>
));
jest.mock('@material-ui/core/Typography', () => ({ children }: any) => (
  <div>{children}</div>
));

describe('<ImageDetails />', () => {
  const props = {
    image: {
      id: '1',
      title: 'Image 1',
      created: 1555853943382,
      details: 'details\nnext line',
      author: 'Rob',
      src: 'https://via.placeholder.com/150',
    },
    onDeleteClick: jest.fn(),
    onEditClick: jest.fn(),
  };

  test('should render and match snapshot with image', () => {
    const wrapper = mount(<ImageDetails {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('should render and match snapshot without image', () => {
    const wrapper = mount(<ImageDetails {...props} image={null} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('should match styles', () => {
    expect(styles(({} as unknown) as Theme)).toMatchSnapshot();
  });
});
