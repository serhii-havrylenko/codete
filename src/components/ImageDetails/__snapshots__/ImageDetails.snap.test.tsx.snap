// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageDetails /> should match styles 1`] = `
Object {
  "actions": Object {
    "display": "flex",
    "justifyContent": "flex-end",
  },
  "avatar": Object {
    "backgroundColor": "#f44336",
  },
  "card": Object {
    "maxWidth": "700px",
  },
  "media": Object {
    "objectFit": "cover",
  },
}
`;

exports[`<ImageDetails /> should render and match snapshot with image 1`] = `
<WithStyles(ImageDetailsUi)
  image={
    Object {
      "author": "Rob",
      "created": 1555853943382,
      "details": "details
next line",
      "id": "1",
      "src": "https://via.placeholder.com/150",
      "title": "Image 1",
    }
  }
  onDeleteClick={[MockFunction]}
  onEditClick={[MockFunction]}
>
  <ImageDetailsUi
    classes={
      Object {
        "actions": "ImageDetailsUi-actions-3",
        "avatar": "ImageDetailsUi-avatar-4",
        "card": "ImageDetailsUi-card-1",
        "media": "ImageDetailsUi-media-2",
      }
    }
    image={
      Object {
        "author": "Rob",
        "created": 1555853943382,
        "details": "details
next line",
        "id": "1",
        "src": "https://via.placeholder.com/150",
        "title": "Image 1",
      }
    }
    onDeleteClick={[MockFunction]}
    onEditClick={[MockFunction]}
  >
    <Component
      className="ImageDetailsUi-card-1"
    >
      <div
        className="ImageDetailsUi-card-1"
      >
        <WithStyles(Component)
          action={
            <Unknown
              data-test-id="edit-button"
              onClick={[Function]}
            >
              <pure(EditIcon) />
            </Unknown>
          }
          avatar={
            <Unknown
              className="ImageDetailsUi-avatar-4"
            >
              Rob
            </Unknown>
          }
          subheader="4/21/2019, 3:39:03 PM"
          title="Image 1"
        >
          <Component
            action={
              <Unknown
                data-test-id="edit-button"
                onClick={[Function]}
              >
                <pure(EditIcon) />
              </Unknown>
            }
            avatar={
              <Unknown
                className="ImageDetailsUi-avatar-4"
              >
                Rob
              </Unknown>
            }
            classes={
              Object {
                "action": "Component-action-5",
              }
            }
            subheader="4/21/2019, 3:39:03 PM"
            title="Image 1"
          >
            <div
              action={
                <Unknown
                  data-test-id="edit-button"
                  onClick={[Function]}
                >
                  <pure(EditIcon) />
                </Unknown>
              }
              avatar={
                <Unknown
                  className="ImageDetailsUi-avatar-4"
                >
                  Rob
                </Unknown>
              }
              classes={
                Object {
                  "action": "Component-action-5",
                }
              }
              subheader="4/21/2019, 3:39:03 PM"
              title="Image 1"
            />
          </Component>
        </WithStyles(Component)>
        <Component
          className="ImageDetailsUi-media-2"
          component="img"
          src="https://via.placeholder.com/150"
          title="Image 1"
        >
          <div />
        </Component>
        <Component>
          <div>
            <Component
              component="pre"
            >
              <div>
                details
next line
              </div>
            </Component>
          </div>
        </Component>
        <Component
          className="ImageDetailsUi-actions-3"
          disableActionSpacing={true}
        >
          <div>
            <Component
              data-test-id="delete-button"
              onClick={[Function]}
            >
              <button
                data-test-id="delete-button"
                onClick={[Function]}
              >
                <pure(DeleteIcon)>
                  <DeleteIcon>
                    <WithStyles(SvgIcon)>
                      <SvgIcon
                        classes={
                          Object {
                            "colorAction": "MuiSvgIcon-colorAction-9",
                            "colorDisabled": "MuiSvgIcon-colorDisabled-11",
                            "colorError": "MuiSvgIcon-colorError-10",
                            "colorPrimary": "MuiSvgIcon-colorPrimary-7",
                            "colorSecondary": "MuiSvgIcon-colorSecondary-8",
                            "fontSizeInherit": "MuiSvgIcon-fontSizeInherit-12",
                            "fontSizeLarge": "MuiSvgIcon-fontSizeLarge-14",
                            "fontSizeSmall": "MuiSvgIcon-fontSizeSmall-13",
                            "root": "MuiSvgIcon-root-6",
                          }
                        }
                        color="inherit"
                        component="svg"
                        fontSize="default"
                        viewBox="0 0 24 24"
                      >
                        <svg
                          aria-hidden="true"
                          className="MuiSvgIcon-root-6"
                          focusable="false"
                          role="presentation"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                          />
                          <path
                            d="M0 0h24v24H0z"
                            fill="none"
                          />
                        </svg>
                      </SvgIcon>
                    </WithStyles(SvgIcon)>
                  </DeleteIcon>
                </pure(DeleteIcon)>
                Delete
              </button>
            </Component>
          </div>
        </Component>
      </div>
    </Component>
  </ImageDetailsUi>
</WithStyles(ImageDetailsUi)>
`;

exports[`<ImageDetails /> should render and match snapshot without image 1`] = `
<WithStyles(ImageDetailsUi)
  image={null}
  onDeleteClick={[MockFunction]}
  onEditClick={[MockFunction]}
>
  <ImageDetailsUi
    classes={
      Object {
        "actions": "ImageDetailsUi-actions-3",
        "avatar": "ImageDetailsUi-avatar-4",
        "card": "ImageDetailsUi-card-1",
        "media": "ImageDetailsUi-media-2",
      }
    }
    image={null}
    onDeleteClick={[MockFunction]}
    onEditClick={[MockFunction]}
  >
    <div>
      Nothing to show. Image is not found
    </div>
  </ImageDetailsUi>
</WithStyles(ImageDetailsUi)>
`;
